Convert `ShallowListEntry` in `InputPattern` handlers

sellout opened the pull request:
## Overview

`ShallowListEntry` as returned by `ls` weren’t being handled by the new `InputPattern` handlers, and so the results of `ls` couldn’t be used as numbered args. This adds support for `ShallowListEntry` to the handlers.

Fixes #5055.

## Interesting/controversial decisions

This additionally makes `hqSplitFromName'` total, which triggers a few other minor adjustments.

Most uses of `ShallowListEntry` in the handler join the entry with the absolute prefix indicating where the `ls` was performed, however there is one handler that only allows relative paths (and is used by `alias.many`). That handler ignores the prefix and provides just the relative `ls` result.

## Test coverage

There is a new transcript that exercises this as exhibited in #5055.

sellout committed:
Remove partiality from `hqSplitFromName'`

This function previously converted the `Name` to a `Path'`, so
recovering the `Name` became partial. It now goes through `Split'`
instead of `Path'`, preserving the guaranteed segment.

sellout committed:
Add a failing transcript for #5055

sellout committed:
Support `ShallowListEntry` in `InputPattern` handlers

These are produced by `ls`, but weren’t included in the handlers, so the
results of `ls` weren’t usable as numbered args.

Fixes #5055.

ceedubs commented on unison-src/transcripts/fix5055.output.md:
````diff
@@ -0,0 +1,93 @@
+```ucm
+.> builtins.merge
````
I tried to leave this comment, but it seems to be lost in the void. It might make sense to create a small static namespace and `ls` it so that this transcript doesn't churn and create noise each time that a builtin is added.

sellout replied:
Fixed. How’s the new one?

sellout committed:
Make the transcript for #5055 more stable

sellout committed:
Ensure `Name`s are absolute after prefixing

In a separate part of the change this PR was extracted from,
`Path.prefixName` ensures the result is absolute, but that is more
involved than other changes here, so this uses the previous technique of
calling `Name.makeAbsolute` afterward.

I don’t think this changes anything important, but the `Name` _is_
always absolute after `Path.prefixName`, so this just ensures that it’s
tracked.

aryairani commented:
Great!
